streams/src/main/java/org/apache/kafka/streams/kstream/internals/StreamsGraphNode.java

```diff
@@ -0,0 +1,174 @@
+                            final boolean repartitionRequired,
+                            final ProcessDetails processDetails,
+                            final String predecessorName) {
+        this.topologyNodeType = topologyNodeType;
+        this.repartitionRequired = repartitionRequired;
+        this.name = name;
+        this.processDetails = processDetails;
+        this.predecessorName = predecessorName;
+    }
+
+    public StreamsGraphNode getPredecessor() {
+        return predecessor;
+    }
+
+
+    public void setPredecessor(StreamsGraphNode predecessor) {
+        Objects.requireNonNull(predecessor, "Can't set a null predecessor");
+        this.predecessor = predecessor;
+        this.predecessorName = predecessor.name();
+    }
+
+
+    public Collection<StreamsGraphNode> getDescendants() {
+        return descendants;
+    }
+
+
+    public void addDescendant(StreamsGraphNode descendant) {
+        descendants.add(descendant);
+    }
+
+
+    public boolean triggersRepartitioning() {
+        return topologyNodeType == MAP || topologyNodeType == GROUP_BY ||
+               topologyNodeType == FLATMAP || topologyNodeType == TRANSFORM || topologyNodeType == SELECT_KEY;
+    }
+
+    public boolean needsRepartitioning() {
+        return repartitionRequired;
+    }
+
+
+    public boolean hasLoggedStore() {
+        return processDetails.getMaterializedInternal() != null && processDetails.getMaterializedInternal().loggingEnabled();
+    }
+
+
+    public void setRepartitionNeeded(boolean needsRepartitioning) {
+        this.repartitionRequired = needsRepartitioning;
+    }
+
+
+    public String name() {
+        return this.name;
+    }
+
+
+    public void setName(String name) {
+        this.name = name;
+    }
+
+    public String getPredecessorName() {
+        return predecessorName;
+    }
+
+    public void setPredecessorName(String predecessorName) {
+        this.predecessorName = predecessorName;
+    }
+
+    public TopologyNodeType getType() {
+        return this.topologyNodeType;
+    }
+
+    public void setProcessed(ProcessDetails processDetails) {
+        this.processDetails = processDetails;
+    }
+
+    public void setId(int id) {
+        if (this.id == null) {
+            this.id = id;
+        }
+    }
+
+    public ProcessDetails getProcessed() {
+        return this.processDetails;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        StreamsGraphNode graphNode = (StreamsGraphNode) o;
+        return topologyNodeType == graphNode.topologyNodeType &&
+               Objects.equals(name, graphNode.name) &&
+               Objects.equals(id, graphNode.id);
+    }
+
+    @Override
+    public int hashCode() {
+
+        return Objects.hash(topologyNodeType, name, id);
+    }
```

Comment:
Can you elaborate on the rationale of these `equals`/`hashcode` overrides?

Since they only contain a few of the properties of this object, there's a possibility of unexpected behavior calling `equal` or adding them to sets and maps. I think the purpose of the override is to support putting these things in sets/maps for some of the optimization logic. If that's the case, another solution would be to create a composite key at that time, and then removing the equals/hashcode override.

This would help to keep this data structure agnostic to the kinds of optimization algorithms that might be applied to it, and also makes the comparison logic more explicit when it is used.